src/infra/core/src/repos/metadata_block_repository_caching_inmem.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::marker::PhantomData;
+
+use async_trait::async_trait;
+use dashmap::DashMap;
+use internal_error::ErrorIntoInternal;
+use kamu_core::{
+    ContainsBlockError,
+    GetBlockError,
+    InsertBlockError,
+    InsertBlockResult,
+    InsertOpts,
+    MetadataBlockRepository,
+    ObjectRepository,
+};
+use opendatafabric::{MetadataBlock, Multihash};
+
+use crate::{MetadataBlockRepositoryExt, MetadataBlockRepositoryImpl};
+
+/////////////////////////////////////////////////////////////////////////////////////////
+// Type shortcuts
+/////////////////////////////////////////////////////////////////////////////////////////
+
+pub type MetadataBlockRepositoryImplWithCache<ObjRepo> =
+    MetadataBlockRepositoryCachingInMem<MetadataBlockRepositoryImpl<ObjRepo>, ObjRepo>;
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+pub struct MetadataBlockRepositoryCachingInMem<WrappedRepo, ObjRepo> {
+    wrapped: WrappedRepo,
+    cache: DashMap<Multihash, MetadataBlock>,
```

Comment:
We discussed lazy computation of deserialized version + storing raw bytes initially.

Reply:
Done, returned back